src/main/bootstrap/cache.rs

```diff
@@ -9,93 +9,194 @@
-    //         try!(self.insert_contacts(contacts, prune));
-    //         // TODO(Team) this implementation is missing and should be considered in next
-    //         // planning
-    //         if Instant::now() > self.last_updated + Self::duration_between_updates() {
-    //             // self.check_bootstrap_contacts();
-    //         }
-    //     }
-
-    //     Ok(())
-    // }
-
-    pub fn read_file(&mut self) -> Vec<PeerInfo> {
-        self.file_handler.read_file().ok().unwrap_or_else(|| vec![])
+    /// Updates cache by reading it from file and returns the current snapshot of peers.
+    pub fn read_file(&self) {
+        let mut inner = self.inner.borrow_mut();
+        inner.peers = inner.file_handler.read_file().unwrap_or_else(|e| {
+            error!("error reading cache file: {}", e);
+            HashSet::new()
+        })
+    }
+
+    /// Inserts given peer to the cache.
+    #[allow(unused)]
+    pub fn put(&self, peer: &PeerInfo) {
+        let mut inner = self.inner.borrow_mut();
+        let _ = inner.peers.insert(peer.clone());
     }
 
-    pub fn remove_peer_acceptor(&mut self, _peer: PeerInfo) {}
-
-    // fn duration_between_updates() -> Duration {
-    //     Duration::from_secs(4 * 60 * 60)
-    // }
-
-    // fn max_contacts() -> usize {
-    //     MAX_BOOTSTRAP_CACHE_CONTACTS
-    // }
-
-    // fn insert_contacts(&mut self,
-    //                    mut contacts: Vec<StaticContactInfo>,
-    //                    prune: Vec<StaticContactInfo>)
-    //                    -> ::Res<()> {
-    //     let mut bootstrap_contacts = self.read_file().unwrap_or_else(|e| {
-    //         debug!("CrustError reading bootstrap cache file: {}.", e);
-    //         Vec::new()
-    //     });
-
-    //     bootstrap_contacts.retain(|contact| !prune.contains(&contact));
-    //     contacts.retain(|contact| !bootstrap_contacts.contains(&contact));
-
-    //     if bootstrap_contacts.is_empty() {
-    //         bootstrap_contacts = contacts;
-    //     } else {
-    //         loop {
-    //             if bootstrap_contacts.len() < Self::max_contacts() && !contacts.is_empty() {
-    //                 bootstrap_contacts.insert(0usize, contacts.remove(0usize));
-    //             } else {
-    //                 break;
-    //             }
-    //         }
-    //     }
-
-    //     Ok(try!(self.file_handler.write_file(&bootstrap_contacts)))
-    // }
+    /// Removes given peer from the cache.
+    pub fn remove(&self, peer: &PeerInfo) {
+        let mut inner = self.inner.borrow_mut();
+        let _ = inner.peers.remove(peer);
+    }
+
+    /// Writes bootstrap cache to disk.
+    pub fn commit(&self) -> ::Res<()> {
+        let inner = self.inner.borrow();
+        inner.file_handler.write_file(&inner.peers)?;
+        Ok(())
+    }
+
+    /// Returns current snapshot of peers in the cache.
+    pub fn peers_vec(&self) -> Vec<PeerInfo> {
+        self.inner.borrow().peers.iter().cloned().collect()
```

Comment:
Better to return it as is, i.e., as an `HashSet` let the user `iter()` over it and collect it the way they like. This will need exposed in `service.rs` because we want the ability to exchange the bootstrap cache file among the peers too (i.e., Authenticator can give it to the Apps and end-users can also manually exchange them out of band).